### PR TITLE
451006-ArthurZhukovski

### PR DESCRIPTION
**Pull request by JonesTwink**
## Comlete version of the tracer.

XML output file:
xml

Console output: 
console

### CODE REVIEW

**Comment by JonesTwink** [comment]
Updated console formatter. Please, leave your review on my pull-request.

File: Tracer/Tracer/Tracer.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+using System.Diagnostics;
+using System.Threading;
+
+namespace Tracer
+{
+    public class Tracer : ITracer
+    {
+
+        public static Tracer Instance { get { return lazy.Value; } }
+        private static readonly Lazy<Tracer> lazy = new Lazy<Tracer>(() => new Tracer());
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/Tracer.cs]: @JonesTwink Thank you for such singleton implementation. 👍

File: Tracer/Tracer/Tracer.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            var parametersAmount = method.GetParameters().Length;
+            var className = method.ReflectedType.Name;
+            var threadId = Thread.CurrentThread.ManagedThreadId;
+            var parentMethod = GetParentMethodName();
+            var newAnalyzedItem = new TraceResultItem(new Stopwatch(), className, method.Name, parametersAmount, threadId, stackFrame.Key, parentMethod);
+            UpdateTraceResult(newAnalyzedItem, threadId);
+            UpdateCallDepth(stackFrame.Key);
+        }
+
+        string GetParentMethodName()
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/Tracer.cs]: @JonesTwink Missing visibility specifier.

**Comment by RamanBut-Husaim** [comment]
@JonesTwink Please, leave a comment when you finish your fixes.

**Comment by JonesTwink** [comment]
@RamanBut-Husaim Thanks for the review. Requested changes has been applied, you may check them out if you wish!

File: Tracer/Tracer/Formatters/XmlTraceResultFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+using System.Xml;
+namespace Tracer.Formatters
+{
+    public  class XmlTraceResultFormatter : ITraceResultFormatter
+    {
+        private string resultFilePath;
+
+        public XmlTraceResultFormatter()
+        {
+            resultFilePath = "TraceResult.xml";
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/Formatters/XmlTraceResultFormatter.cs]: @JonesTwink Mmm, `XmlTraceResultFormatter` should accept file path as a parameter in the constructor. :)

File: Tracer/Tracer/Formatters/ConsoleTraceResultFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                if (depth == 1)
+                    Console.WriteLine("\nПоток №{0}",analyzedItem.threadId);
+                Console.WriteLine("{0}|\n{0}{1}",addTabs(depth), analyzedItem.ToString());
+                if (analyzedItem.children != null)
+                    Print(analyzedItem.children, depth);
+            }
+            
+        }
+
+        private string addTabs(int n)
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: @JonesTwink Method naming. `PascalCase`

File: Tracer/Tracer/Formatters/XmlTraceResultFormatter.cs (hunk 3 above)
Comment (JonesTwink) [inline Tracer/Tracer/Formatters/XmlTraceResultFormatter.cs]: @RamanBut-Husaim  But there's another constructor which accepts  filepath as a parameter right below this one. And this parameterless constructor  has been added in order to provide higher versatility as it sets up a default filepath value (for instance, when a programmer doesn't want to specify his custom filepath). Should I still  remove it?

File: Tracer/Tracer/Formatters/ConsoleTraceResultFormatter.cs (hunk 4 above)
Comment (JonesTwink) [inline Tracer/Tracer/Formatters/ConsoleTraceResultFormatter.cs]: What a miss... Shame on me (:

File: Tracer/Tracer/Formatters/XmlTraceResultFormatter.cs (hunk 3 above)
Comment (RamanBut-Husaim) [inline Tracer/Tracer/Formatters/XmlTraceResultFormatter.cs]: @JonesTwink Yep. You could use this string value as the default parameter value in the constructor with file path. It is described [here](https://www.dotnetperls.com/optional-parameters).

Comment (JonesTwink) [inline Tracer/Tracer/Formatters/XmlTraceResultFormatter.cs]: @RamanBut-Husaim  Thanks for advice! This option has somehow flown out of my head.

File: Tracer/Tracer/TraceResultItem.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            this.threadId = threadId;
+            this.callDepth = callDepth;
+            this.parent = parent;
+            try
+            {
+                timer.Start();
+            }
+            catch (NullReferenceException)
+            {
+                throw new NullReferenceException("Timer is missing! Check arguments passed to TraceResultItem.");
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer/TraceResultItem.cs]: @JonesTwink Do not throw `NullReferenceException` from your code. This type of the exception show the bug in your program. You could replace it with `InvalidOperationException` or smth else.

**Comment by JonesTwink** [comment]
@RamanBut-Husaim
> replace it with `InvalidOperationException` or smth else. Z
Done.

**Comment by RamanBut-Husaim** [comment]
@JonesTwink Accepted.

**Pull request merged**